enterprise/com/src/main/java/org/neo4j/com/storecopy/FileMoveProvider.java

```diff
@@ -0,0 +1,134 @@
+    {
+        this.pageCache = pageCache;
+        this.fileMoveActionInformer = fileMoveActionInformer;
+    }
+
+    /**
+     * Construct a stream of files that are to be moved
+     *
+     * @param dir the source location of the move action
+     * @return a stream of the entire contents of the source location that can be applied to a target location to perform a move
+     */
+    public Stream<FileMoveAction> traverseGenerateMoveActions( File dir )
+    {
+        return traverseGenerateMoveActions( dir, dir );
+    }
+
+    /**
+     * Copies <b>the contents</b> from the directory to the base target path.
+     * <p>
+     * This is confusing, best read the tests
+     *
+     * @param dir this directory and all the child paths under it are subject to move
+     * @param basePath this is the parent of your intended target directory.
+     * @return a stream of individual move actions which can be iterated and applied whenever
+     */
+    public Stream<FileMoveAction> traverseGenerateMoveActions( File dir, File basePath )
+    {
+        // Note that flatMap is an *intermediate operation* and therefor always lazy.
+        // It is very important that the stream we return only *lazily* calls out to expandTraverseFiles!
+        return Stream.of( dir ).flatMap( d -> expandTraverseFiles( d, basePath ) );
+    }
+
+    private Stream<FileMoveAction> expandTraverseFiles( File dir, File basePath )
+    {
+        List<File> listing = listFiles( dir );
+        if ( listing.isEmpty() )
+        {
+            return Stream.of( moveFileCorrectly( dir, basePath ) );
+        }
+        Stream<File> files = listing.stream().filter( this::isFile );
+        Stream<File> dirs = listing.stream().filter( this::isDirectory );
+        Stream<FileMoveAction> moveFiles = files.map( f -> moveFileCorrectly( f, basePath ) );
+        Stream<FileMoveAction> traverseDirectories = dirs.flatMap( d -> traverseGenerateMoveActions( d, basePath ) );
+        return Stream.concat( moveFiles, traverseDirectories );
+    }
+
+    private boolean isFile( File file )
+    {
+        if ( fileMoveActionInformer.shouldBeManagedByPageCache( file.getName() ) )
+        {
+            return !pageCache.getCachedFileSystem().isDirectory( file );
+        }
+        return file.isFile();
+    }
+
+    private boolean isDirectory( File file )
+    {
+        if ( fileMoveActionInformer.shouldBeManagedByPageCache( file.getName() ) )
+        {
+            return pageCache.getCachedFileSystem().isDirectory( file );
+        }
+        return file.isDirectory();
+    }
+
+    private List<File> listFiles( File dir )
+    {
+        List<File> pageCacheFiles = asList( safeArray( pageCache.getCachedFileSystem().listFiles( dir ) ) );
+        List<File> fsFiles = safeArray( dir.listFiles() );
+        return Stream.of( pageCacheFiles, fsFiles ).flatMap( List::stream ).collect( Collectors.toList() );
+    }
+
+    private List<File> safeArray( File[] files )
```

Comment:
`nullSafeArray` may be better name? There may also be an appropriate place for this in `community/common`.